fix: Editing selected text fails to apply changes across multiple blo…

syuzanvs opened the pull request:
…ck elements

- Handle selected content when processing slash commands
  - Added logic to retrieve and convert selected content to HTML when a command is executed.
  - Ensured that the command and selected content are properly assigned to variables for further processing.

- Improve AI tag insertion logic in fetchAndProcessGptResponse method
  - Added logic to handle insertion of 'ai-tag' elements for both inline and block-level content.
  - Ensured proper positioning of 'ai-tag' elements based on the current selection.
  - Improved handling of nested elements and parent-child relationships during tag insertion.

- Enhance updateContent method to handle inline and block-level content updates
   - Added logic to handle text content separately from HTML content.
   - Improved handling of inline elements by checking for list elements and updating text content accordingly.
   - Ensured proper insertion of model fragments for block-level content updates.

- Refactor clearParentContent method to handle empty parent elements
   - Added logic to remove parent elements if they become empty after removing a range.
   - Ensured that 'ai-tag' elements are also removed if they are the only child and not a root element.
   - Improved handling of inline-slash elements to check if the line is empty before removal.

- Refine AI tag cleanup logic in processCompleted method
  - Improved regular expressions to accurately remove 'ai-tag' elements and their placeholders.
  - Ensured that both inline and block-level 'ai-tag' elements are properly cleaned up from the editor content.

- Updated the CSS selectors for the .ck ai-tag class to ensure proper styling and functionality.
- Corrected the selector syntax to ensure that the styles apply correctly to the AI tag elements.
- Removed unnecessary spaces in the selectors to adhere to best practices and improve readability.
- "Replace" is changed with "Return" so that it won't return the whole line but only the selected text, as it reads the whole line but we don't neet it to return the whole line.

Closes #80

jjroelofs reviewed (changes requested):
## Help me understand

This is a complex PR, I rely on the commit message to understand it and learn what engineering decisions you made and why. Currently the commit message is not correct or clear.

## Blue artifacts

During normal AI writing I see blue artifacts and newlines before and after the content, can we get rid of these?

![image](https://github.com/user-attachments/assets/9414b6b6-763e-43f9-b5da-f7bf3a379896)

## Prompt change

I see the prompt was changed: `Replace` > `Return`  
Please explain this change in the commit message. If we keep the change please also update the README:

https://github.com/dxpr/ckeditor5-ai-agent?tab=readme-ov-file#context-requirements

Changes to the prompt should be made carefully as updating the prompt constitutes a backward-incompatible change in the Drupal module I'm working on (we need to update the prompt component in the database, create a migration for this, release notes, etc.).

## Merge conflicts

Please fix them

jjroelofs reviewed (changes requested):
It still seems to me like many parts of the first commit message in this PR do not reflect the code changes:
```
- Added methods to handle AI Agent button functionality, including dropdown menu and command execution.
- Implemented placeholder management to show/hide placeholders based on editor content.
- Added loader element to indicate processing and methods to show/hide the loader.
- Included error tooltip element for displaying error messages with methods to show/hide the tooltip.
- Registered schema and conversions for custom tags (inline-slash and ai-tag).
- Attached event listeners for handling user interactions and content changes.
- Enhanced AI Agent Service with moderation, content processing, and error handling.
- Improved placeholder positioning for RTL languages.
```


Please look at this again critically and squash your commits into a single atomic commit with clear documentation, like in your last commit

jjroelofs reviewed (changes requested):
The latest update causes a regression when editing all text in the editor:

![Untitled Video January 7, 2025 2_14 PM](https://github.com/user-attachments/assets/5ccaf0e4-da95-424c-b8ec-e56eb8e49fda)


```
ERROR
model-nodelist-offset-out-of-bounds {"offset":4,"nodeList":[{"attributes":{"id":"ai-1736255678338-inline"},"name":"ai-tag"}]}
Read more: https://ckeditor.com/docs/ckeditor5/latest/support/error-codes.html#error-model-nodelist-offset-out-of-bounds
CKEditorError: model-nodelist-offset-out-of-bounds {"offset":4,"nodeList":[{"attributes":{"id":"ai-1736255678338-inline"},"name":"ai-tag"}]}
Read more: https://ckeditor.com/docs/ckeditor5/latest/support/error-codes.html#error-model-nodelist-offset-out-of-bounds
    at NodeList.offsetToIndex (webpack://@dxpr/ckeditor5-ai-agent/./node_modules/@ckeditor/ckeditor5-engine/src/model/nodelist.js?:133:19)
    at Element.offsetToIndex (webpack://@dxpr/ckeditor5-ai-agent/./node_modules/@ckeditor/ckeditor5-engine/src/model/element.js?:131:31)
    at getNodeBeforePosition (webpack://@dxpr/ckeditor5-ai-agent/./node_modules/@ckeditor/ckeditor5-engine/src/model/position.js?:927:51)
    at get nodeBefore (webpack://@dxpr/ckeditor5-ai-agent/./node_modules/@ckeditor/ckeditor5-engine/src/model/position.js?:167:16)
    at Range.getContainedElement (webpack://@dxpr/ckeditor5-ai-agent/./node_modules/@ckeditor/ckeditor5-engine/src/model/range.js?:482:40)
    at TableUtils.getSelectedTableCells (webpack://@dxpr/ckeditor5-ai-agent/./node_modules/@ckeditor/ckeditor5-table/src/tableutils.js?:734:35)
    at TableSelection._handleDeleteContent (webpack://@dxpr/ckeditor5-ai-agent/./node_modules/@ckeditor/ckeditor5-table/src/tableselection.js?:225:47)
    at listenTo.priority (webpack://@dxpr/ckeditor5-ai-agent/./node_modules/@ckeditor/ckeditor5-table/src/tableselection.js?:54:67)
    at Model.fire (webpack://@dxpr/ckeditor5-ai-agent/./node_modules/@ckeditor/ckeditor5-utils/src/emittermixin.js?:159:47)
    at <computed> [as deleteContent] (webpack://@dxpr/ckeditor5-ai-agent/./node_modules/@ckeditor/ckeditor5-utils/src/observablemixin.js?:191:29)
```

jjroelofs reviewed (changes requested):
As was pointed out in 2 prior reviews already, the commit log still shows incorrect information:
```
- Added methods to handle AI Agent button functionality, including dropdown menu and command execution.
- Implemented placeholder management to show/hide placeholders based on editor content.
- Added loader element to indicate processing and methods to show/hide the loader.
- Included error tooltip element for displaying error messages with methods to show/hide the tooltip.
- Registered schema and conversions for custom tags (inline-slash and ai-tag).
- Attached event listeners for handling user interactions and content changes.
- Enhanced AI Agent Service with moderation, content processing, and error handling.
- Improved placeholder positioning for RTL languages.
```

Please clean up history using a tool like Git Squash. The list of changes in the PR description looks good, I suggest you use it as a base for creating a final atomic commit.

syuzanvs committed:
fix: Editing selected text fails to apply changes across multiple block elements

- Handle selected content when processing slash commands
 - Added logic to retrieve and convert selected content to HTML when a command is executed.
 - Ensured that the command and selected content are properly assigned to variables for further processing.

- Improve AI tag insertion logic in fetchAndProcessGptResponse method
 - Added logic to handle insertion of 'ai-tag' elements for both inline and block-level content.
 - Ensured proper positioning of 'ai-tag' elements based on the current selection.
 - Improved handling of nested elements and parent-child relationships during tag insertion.

- Enhance updateContent method to handle inline and block-level content updates
 - Added logic to handle text content separately from HTML content.
 - Improved handling of inline elements by checking for list elements and updating text content accordingly.
 - Ensured proper insertion of model fragments for block-level content updates.

- Refactor clearParentContent method to handle empty parent elements
 - Added logic to remove parent elements if they become empty after removing a range.
 - Ensured that 'ai-tag' elements are also removed if they are the only child and not a root element.
 - Improved handling of inline-slash elements to check if the line is empty before removal.

- Refine AI tag cleanup logic in processCompleted method
 - Improved regular expressions to accurately remove 'ai-tag' elements and their placeholders.
 - Ensured that both inline and block-level 'ai-tag' elements are properly cleaned up from the editor content.

- Updated the CSS selectors for the .ck ai-tag class to ensure proper styling and functionality.

- Corrected the selector syntax to ensure that the styles apply correctly to the AI tag elements.

- Removed unnecessary spaces in the selectors to adhere to best practices and improve readability.

- "Replace" is changed with "Return" so that it won't return the whole line but only the selected text, as it reads the whole line but we don't neet it to return the whole line.

Closes #80

jjroelofs reviewed (approved):
great work!
